File: app/src/main/java/com/marcelorcorrea/falae/fragment/PageFragment.java
package com.marcelorcorrea.falae.fragment;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;
import android.widget.ImageView;

import com.marcelorcorrea.falae.R;
import com.marcelorcorrea.falae.adapter.ItemPagerAdapter;
import com.marcelorcorrea.falae.model.Page;

import static com.marcelorcorrea.falae.R.id.pager;


public class PageFragment extends Fragment {
    private static final String PAGE_PARAM = "pageParam";

    private Page page;
    private OnFragmentInteractionListener mListener;
    private ViewPager mPager;
    private ItemPagerAdapter mPagerAdapter;
    private ImageView leftNav;
    private ImageView rightNav;
    private FrameLayout leftNavHolder;
    private FrameLayout rightNavHolder;

    public PageFragment() {
    }

    public static PageFragment newInstance(Page page) {
        PageFragment fragment = new PageFragment();
        Bundle args = new Bundle();
        args.putParcelable(PAGE_PARAM, page);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            page = getArguments().getParcelable(PAGE_PARAM);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_page, container, false);
        leftNav = (ImageView) view.findViewById(R.id.left_nav);
        rightNav = (ImageView) view.findViewById(R.id.right_nav);

        leftNavHolder = (FrameLayout) view.findViewById(R.id.left_nav_holder);
        rightNavHolder = (FrameLayout) view.findViewById(R.id.right_nav_holder);

        ViewTreeObserver vto = view.getViewTreeObserver();
        vto.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
                    view.getViewTreeObserver().removeGlobalOnLayoutListener(this);
                } else {
                    view.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                }
                int leftNavMeasuredWidth = leftNavHolder.getMeasuredWidth();
                int rightNavMeasureWidth = rightNavHolder.getMeasuredWidth();

                mPager = (ViewPager) view.findViewById(pager);
                mPagerAdapter = new ItemPagerAdapter(getChildFragmentManager(), page, leftNavMeasuredWidth + rightNavMeasureWidth);
                mPager.setAdapter(mPagerAdapter);

                ViewGroup.MarginLayoutParams pagerLayoutParams = (ViewGroup.MarginLayoutParams) mPager.getLayoutParams();
                pagerLayoutParams.leftMargin += leftNavMeasuredWidth;
                pagerLayoutParams.rightMargin += rightNavMeasureWidth;

                mPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
                    @Override
                    public void onPageSelected(int position) {
                        handleNavButtons();
                    }
                });

                if (shouldEnableNavButtons()) {
                    handleNavButtons();
                }
                leftNavHolder.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        int tab = mPager.getCurrentItem();
                        if (tab > 0) {
                            tab--;
                            speak(getString(R.string.previous));
                            mPager.setCurrentItem(tab);
                        } else if (tab == 0) {
                            mPager.setCurrentItem(tab);
                        }
                    }
                });
                rightNavHolder.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        int tab = mPager.getCurrentItem();
                        if (mPagerAdapter.getCount() > 1 && tab != mPagerAdapter.getCount() - 1) {
                            speak(getString(R.string.next));
                        }
                        tab++;
                        mPager.setCurrentItem(tab);
                    }
                });
            }
        });

        return view;
    }

    private void handleNavButtons() {
        enableNavButtons();
        if (shouldDisableLeftNavButton()) {
            leftNav.setVisibility(View.INVISIBLE);
        }
        if (shouldDisableRightButton()) {
            rightNav.setVisibility(View.INVISIBLE);
        }
    }

    private void enableNavButtons() {
        leftNav.setVisibility(View.VISIBLE);
        rightNav.setVisibility(View.VISIBLE);
    }

    private boolean shouldEnableNavButtons() {
        return mPagerAdapter.getCount() > 1;
    }

    private boolean shouldDisableLeftNavButton() {
        return mPager.getCurrentItem() == 0;
    }

    private boolean shouldDisableRightButton() {
        return mPager.getCurrentItem() >= mPagerAdapter.getCount() - 1;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void speak(String msg) {
        mListener.speak(msg);
    }

    public interface OnFragmentInteractionListener {
        void speak(String msg);
    }
}

File: app/src/main/java/com/marcelorcorrea/falae/fragment/PageFragment.kt
package com.marcelorcorrea.falae.fragment

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.FrameLayout
import android.widget.ImageView
import com.marcelorcorrea.falae.R
import com.marcelorcorrea.falae.R.id.pager
import com.marcelorcorrea.falae.adapter.ItemPagerAdapter
import com.marcelorcorrea.falae.model.Page


class PageFragment : Fragment() {

    private var page: Page? = null
    private lateinit var mListener: OnFragmentInteractionListener
    private lateinit var mPager: ViewPager
    private lateinit var mPagerAdapter: ItemPagerAdapter
    private lateinit var leftNav: ImageView
    private lateinit var rightNav: ImageView
    private lateinit var leftNavHolder: FrameLayout
    private lateinit var rightNavHolder: FrameLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            page = arguments.getParcelable(PAGE_PARAM)
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_page, container, false)
        leftNav = view.findViewById(R.id.left_nav) as ImageView
        rightNav = view.findViewById(R.id.right_nav) as ImageView
        leftNavHolder = view.findViewById(R.id.left_nav_holder) as FrameLayout
        rightNavHolder = view.findViewById(R.id.right_nav_holder) as FrameLayout

        val vto = view.viewTreeObserver
        vto.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
                    view.viewTreeObserver.removeGlobalOnLayoutListener(this)
                } else {
                    view.viewTreeObserver.removeOnGlobalLayoutListener(this)
                }
                mPager = view.findViewById(pager) as ViewPager
                val navHoldersSize = java.lang.Double.valueOf(mPager.measuredWidth * 0.065)!!.toInt()
                leftNav.layoutParams.width = navHoldersSize
                leftNav.layoutParams.height = navHoldersSize
                rightNav.layoutParams.width = navHoldersSize
                rightNav.layoutParams.height = navHoldersSize
                leftNavHolder.layoutParams.width = navHoldersSize
                rightNavHolder.layoutParams.width = navHoldersSize

                mPagerAdapter = ItemPagerAdapter(childFragmentManager, page!!, navHoldersSize * 2)
                mPager.adapter = mPagerAdapter

                val pagerLayoutParams = mPager.layoutParams as ViewGroup.MarginLayoutParams
                pagerLayoutParams.leftMargin += navHoldersSize
                pagerLayoutParams.rightMargin += navHoldersSize
                mPager.addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
                    override fun onPageSelected(position: Int) {
                        handleNavButtons()
                    }
                })

                if (shouldEnableNavButtons()) {
                    handleNavButtons()
                }
                leftNavHolder.setOnClickListener {
                    var tab = mPager.currentItem
                    if (tab > 0) {
                        tab--
                        speak(getString(R.string.previous))
                        mPager.currentItem = tab
                    } else if (tab == 0) {
                        mPager.currentItem = tab
                    }
                }
                rightNavHolder.setOnClickListener {
                    var tab = mPager.currentItem
                    if (mPagerAdapter.count > 1 && tab != mPagerAdapter.count - 1) {
                        speak(getString(R.string.next))
                    }
                    tab++
                    mPager.currentItem = tab
                }
            }
        })

        return view
    }

    private fun handleNavButtons() {
        enableNavButtons()
        if (shouldDisableLeftNavButton()) {
            leftNav.visibility = View.INVISIBLE
        }
        if (shouldDisableRightButton()) {
            rightNav.visibility = View.INVISIBLE
        }
    }

    private fun enableNavButtons() {
        leftNav.visibility = View.VISIBLE
        rightNav.visibility = View.VISIBLE
    }

    private fun shouldEnableNavButtons(): Boolean = mPagerAdapter.count > 1

    private fun shouldDisableLeftNavButton(): Boolean = mPager!!.currentItem == 0

    private fun shouldDisableRightButton(): Boolean = mPager.currentItem >= mPagerAdapter.count - 1

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
    }

    fun speak(msg: String) {
        mListener.speak(msg)
    }

    interface OnFragmentInteractionListener {
        fun speak(msg: String)
    }

    companion object {
        private val PAGE_PARAM = "pageParam"

        fun newInstance(page: Page): PageFragment {
            val fragment = PageFragment()
            val args = Bundle()
            args.putParcelable(PAGE_PARAM, page)
            fragment.arguments = args
            return fragment
        }
    }
}
